keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,360 @@
+  aggregated APIs without further burdening system administrators or Kubernetes
+  distributions.
+
+## Proposal
+
+### Alpha workflow
+
+We will create a **kube-migrator** command line tool. At the alpha stage,
+**kube-migrator** needs to be manually invoked, and does not handle custom
+resources. 
+
+After all the apiservers are at the desired version, including kube-apiservers
+and aggregated apiservers, the cluster administrator runs `kube-migrator
+install` command. It
+* deletes any existing `deployment` of **kube-migrator controller** 
+* creates a `deployment` of **kube-migrator controller** 
+* uses the same kubeconfig file as **kubectl** does.
+
+On start-up, the **kube-migrator controller** does the following,
+* discovers all supported APIs.
+* lists all custom resource definitions, excludes them from the API list
+  retrieved in the previous step.
+* for each non-custom resource, creates a `migration` CR (see the [API
+  section][] for the schema) if it doesn't exist already.
+* for each newly created `migration` instance, increments the
+  `migration.spec.generation`, and logs the reason.
+* the `ownerReferences` of the `migration` objects are set to the **kube-migrator
+  controller** `deployment`. Thus, when user invokes the **kube-migrator** tool,
+  the old `migration`s are deleted with the old `deployment`.
+
+The control loop of **kube-migrator controller** does the following:
+* watches the instances of the `migration` CR.
+* if `migration.spec.generation != migration.status.observedGeneration`
+    * deletes existing **migration worker** `job` for the resource type.
+    * launches new **migration worker** `job` to migrate the resource type. The
+      `job` is annotated with the generation.
+    * makes sure there is only one **migration worker** `job` running at a time,
+      to avoid overloading the apiserver.
+    * updates `migration.status.observedGeneration`.
+* watches the **migration worker** jobs. Updates
+  `migration.status.completedGeneration` when a job completes.
+* short-circuit: if `migration.spec.groupVersions` shows that the resource has
+  only one version, then its `.status.observedGeneration` and
+  `.status.completedGeneration` are updated directly without launching the
+  worker `job`.
+
+The **migrator worker** `job` runs the equivalence of `oc adm migrate storage
+--include=<resource type>` to migrate a resource type. The **migrator worker**
+uses API chunking to retrieve partial lists of a resource type and thus can
+migrate a small chunk at a time. It stores the [continue token] in a `configmap`
+to preserve progress. With the inconsistent continue token introduced in
+[#67284][], the **migrator worker** does not need to worry about expired continue
+token.
+
+[#67284]:https://github.com/kubernetes/kubernetes/pull/67284
+[API section]:#api
+
+The cluster admin can run the `kube-migrator progress` command to check if the
```

Comment:
so kube-migrator has 'install' and 'progress' commands, what about 'uninstall' that can clean up everything related to the migration.

Reply:
Yeah, we can add such a subcommand, would be trivial. I will skip it in the KEP.

Reply:
Can this state be kept in a CRD, such that I can use kubectl to read the current progress? It seems contra our general direction to introduce new command line tools.

Reply:
Removed the command, user can just run `kubectl get migrations -o jsonpath="{.status.completed}"`.